### PR TITLE
Scheduler.py typing - remove `allow_incomplete_defs`

### PR DESCRIPTION
**Pull request by fjetter**
This removes the flag for `allow_incomplete_defs` from scheduler.py. The step for `allow_untyped_defs` is rather minimal and concerns mostly `Scheduler.__init__` but I stopped and leave this to another PR.

I do not intend do engage on a thorough review/feedback cycle unless something is horribly wrong. This is supposed to be a quick drive by to enable this flag. We can correct inaccurate types in follow up iterations. Spending too much time on this being open just increases risk of having merge conflicts, i.e. I would prefer addressing *any* review comments in a follow up

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -4107,20 +4108,21 @@ async def add_worker(
...
-        pid=0,
-        services=None,
-        local_directory=None,
-        versions: dict[str, Any] | None = None,
-        nanny=None,
-        extra=None,
+        nthreads: int,
+        name: str,
+        resolve_address: bool = True,
+        now: int,
```
Comment (crusaderky) [inline distributed/scheduler.py:4114]:
```suggestion
        now: float,
```

File: distributed/scheduler.py [hunk 2]
```diff
@@ -4107,20 +4108,21 @@ async def add_worker(
...
+        resources: dict,
```
Comment (crusaderky) [inline distributed/scheduler.py:4115]:
```suggestion
        resources: dict[str, float],
```

File: distributed/scheduler.py [hunk 3]
```diff
@@ -4107,20 +4108,21 @@ async def add_worker(
...
+        # FIXME: This is never submitted by the worker
+        host_info: Any = None,
```
Comment (crusaderky) [inline distributed/scheduler.py:4117]:
```suggestion
        host_info: None = None,
```

File: distributed/scheduler.py [hunk 4]
```diff
@@ -4107,20 +4108,21 @@ async def add_worker(
...
+        memory_limit: int | None,
+        metrics: dict,
```
Comment (crusaderky) [inline distributed/scheduler.py:4119]:
```suggestion
        metrics: dict[str, Any],
```

File: distributed/scheduler.py [hunk 5]
```diff
@@ -7107,20 +7124,21 @@ def get_has_what(self, workers=None):
         else:
             return {w: [ts.key for ts in ws.has_what] for w, ws in self.workers.items()}
 
-    def get_ncores(self, workers=None):
+    def get_ncores(self, workers: Iterable[str] | None = None) -> dict:
```
Comment (crusaderky) [inline distributed/scheduler.py:7127]:
```suggestion
    def get_ncores(self, workers: Iterable[str] | None = None) -> dict[str, int]:
```

File: distributed/scheduler.py [hunk 6]
```diff
@@ -7095,7 +7112,7 @@ def get_who_has(self, keys: Iterable[str] | None = None) -> dict[str, list[str]]
                 key: [ws.address for ws in ts.who_has] for key, ts in self.tasks.items()
             }
 
-    def get_has_what(self, workers=None):
+    def get_has_what(self, workers: Iterable[str] | None = None) -> dict:
```
Comment (crusaderky) [inline distributed/scheduler.py:7115]:
```suggestion
    def get_has_what(self, workers: Iterable[str] | None = None) -> dict[str, list[str]]:
```

File: distributed/scheduler.py [hunk 7]
```diff
@@ -7065,15 +7082,15 @@ def log_worker_event(
...
+    def subscribe_worker_status(self, comm: Comm) -> str:
         WorkerStatusPlugin(self, comm)
         ident = self.identity()
         for v in ident["workers"].values():
             del v["metrics"]
             del v["last_seen"]
         return ident
 
-    def get_processing(self, workers=None):
+    def get_processing(self, workers: Iterable[str] | None = None) -> dict:
```
Comment (crusaderky) [inline distributed/scheduler.py:7093]:
```suggestion
    def get_processing(self, workers: Iterable[str] | None = None) -> dict[str, list[str]]:
```

File: distributed/scheduler.py [hunk 8]
```diff
@@ -7107,20 +7124,21 @@ def get_has_what(self, workers=None):
...
-    def get_ncores(self, workers=None):
+    def get_ncores(self, workers: Iterable[str] | None = None) -> dict:
         if workers is not None:
             workers = map(self.coerce_address, workers)
             return {w: self.workers[w].nthreads for w in workers if w in self.workers}
         else:
             return {w: ws.nthreads for w, ws in self.workers.items()}
 
-    def get_ncores_running(self, workers=None):
+    def get_ncores_running(self, workers: Iterable[str] | None = None) -> dict:
```
Comment (crusaderky) [inline distributed/scheduler.py:7134]:
```suggestion
    def get_ncores_running(self, workers: Iterable[str] | None = None) -> dict[str, int]:
```

File: distributed/scheduler.py [hunk 9]
```diff
@@ -7204,21 +7224,30 @@ async def benchmark_hardware(self) -> "dict[str, dict[str, float]]":
         return result
 
     @log_errors
-    def get_nbytes(self, keys=None, summary=True):
+    def get_nbytes(
+        self, keys: Iterable[str] | None = None, summary: bool = True
+    ) -> dict:
```
Comment (crusaderky) [inline distributed/scheduler.py:7229]:
```suggestion
    ) -> dict[str, int]:
```

File: distributed/scheduler.py [hunk 10]
```diff
@@ -7284,26 +7313,31 @@ def get_task_prefix_states(self):
 
         return state
 
-    def get_task_status(self, keys=None):
+    def get_task_status(self, keys: Iterable[str]) -> dict:
```
Comment (crusaderky) [inline distributed/scheduler.py:7316]:
```suggestion
    def get_task_status(self, keys: Iterable[str]) -> dict[str, TaskStateState | None]:
```

File: distributed/scheduler.py [hunk 11]
```diff
@@ -7450,15 +7486,15 @@ def add_resources(self, worker: str, resources=None):
...
-    def remove_resources(self, worker):
+    def remove_resources(self, worker: str) -> None:
         ws: WorkerState = self.workers[worker]
         for resource in ws.resources:
-            dr: dict = self.resources.get(resource, None)
-            if dr is None:
-                self.resources[resource] = dr = {}
+            if resource not in self.resources:
+                self.resources[resource] = {}
+            dr: dict = self.resources[resource]
```
Comment (crusaderky) [inline distributed/scheduler.py:7494]:
```suggestion
            dr = self.resources.setdefault(resource, {})
```

**Review by crusaderky — CHANGES_REQUESTED** [review]

**Comment by crusaderky** [comment]
Some nits only. It should be just a matter of committing all suggestions and we're ready to go.

**Review by crusaderky — APPROVED** [review]
LGTM as soon as CI turns out green

**Comment by graingert** [comment]
github actions is down at the moment https://www.githubstatus.com/incidents/z3c6q056q332 so CI looks green but the mypy check didn't run

**Pull request merged**